### PR TITLE
Improvements in handling of rorc-shr pairs

### PR DESCRIPTION
**Pull request by uxmal**
A multiword right-shift can be computed by performing an initial `shr` operation, and then cascading the resulting carry out bit across one or more `rorc/rcr` operations.

### CODE REVIEW

**Comment by uxmal** [comment]
Fascinating: failures on linux_arm64... @smx-smx, any comments on how to troubleshoot?

File: src/Decompiler/Evaluation/ExpressionSimplifier.cs [hunk 1]
```diff
@@ -1163,6 +1172,33 @@ public virtual (Expression, bool) VisitMkSequence(MkSequence seq)
...
+                    slice.Expression is BinaryExpression bin && 
+                    bin.Operator is IMulOperator)
+                {
+                    if (mul is null)
+                        mul = bin;
+                    else if (!cmp.Equals(mul.Left, bin.Left) ||
+                            !cmp.Equals(mul.Right, bin.Right))
+                        return null;
+                    else
+                        mul = bin;
```
Comment (ptomin) [inline src/Decompiler/Evaluation/ExpressionSimplifier.cs:1190]: Looks like incorrect logic for me. If sequence contains any slice with multiplication operator then this code discards other elements of sequence.

File: src/Decompiler/Analysis/ConditionCodeEliminator.cs [hunk 2]
```diff
@@ -453,30 +471,45 @@ private Instruction TransformRorC(Application rorc, Assignment a)
...
             var block = sidOrigLo.DefStatement!.Block;
             var expShrSrc = shift.Left;
             var expRorSrc = rorc.Arguments[0];
 
-            var stmGrf = sidGrf!.DefStatement;
-            block.Statements.Remove(stmGrf!);
+            foreach (var sid in sidsToKill)
+            {
+                ssa.DeleteStatement(sid.DefStatement!);
+                ssa.Identifiers.Remove(sid);
```
Comment (ptomin) [inline src/Decompiler/Analysis/ConditionCodeEliminator.cs:495]: What if `sid` has other uses?

**Review by ptomin — CHANGES_REQUESTED** [review]
John, I have added comments.

**Comment by smx-smx** [comment]
> Fascinating: failures on linux_arm64... @smx-smx, any comments on how to troubleshoot?

The VM has 4GB, and we're running Out Of Memory

image

**Comment by uxmal** [comment]
Indeed there is a big performance regression.  I'm seeing runtime being doubled.

**Comment by ptomin** [comment]
> Indeed there is a big performance regression. I'm seeing runtime being doubled.
Validation `SSA` can reduce performance. See https://github.com/uxmal/reko/pull/1111#discussion_r774628731

File: src/Decompiler/Analysis/LongAddRewriter.cs [hunk 3]
```diff
@@ -287,16 +285,21 @@ public void ReplaceLongAdditions(Block block)
...
+                if (cond is null)
                     continue;
+
                 var hiInstr = FindUsingInstruction(block, cond.FlagGroup, loInstr);
-                if (hiInstr == null)
+                if (hiInstr is null)
                     continue;
-
+                if (hiInstr.Statement?.ToString() == "ah_29 = ah_27 + C_28")
+                    _ = hiInstr.ToString();//$DEBUG
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Decompiler/Analysis/LongAddRewriter.cs:298]: Removed

File: src/Decompiler/Analysis/LongAddRewriter.cs [hunk 4]
```diff
@@ -287,16 +285,21 @@ public void ReplaceLongAdditions(Block block)
...
+                ssa.Validate(s => { });
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Decompiler/Analysis/LongAddRewriter.cs:299]: Removed

File: src/Decompiler/Analysis/LongAddRewriter.cs [hunk 5]
```diff
@@ -287,16 +285,21 @@ public void ReplaceLongAdditions(Block block)
...
+                trace.Verbose("Larw: {0}: found add/sub pair {1} / {2}", block.DisplayName, loInstr.Statement!, hiInstr.Statement!);
                 CreateLongInstruction(loInstr, hiInstr);
+                ssa.Validate(s => { Console.WriteLine("Larw: {0}", s); });
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Decompiler/Analysis/LongAddRewriter.cs:302]: Removed

File: src/UnitTests/Analysis/LongAddRewriterTests.cs [hunk 6]
```diff
@@ -419,5 +419,129 @@ public void Larw_do_not_span_multiple_blocks()
...
+
+                m.Assign(ax, m.IAdd(ax, m.Mem16(m.IAdd(bx, 6))));
+                m.Assign(SCZ, m.Cond(ax));
+                //m.Alias(CF, m.Slice(PrimitiveType.Bool, SCZO, 1));
+                m.Assign(dx, m.IAdd(m.IAdd(dx, m.Mem16(m.IAdd(bx, 8))), CF));
+                m.Assign(SCZ, m.Cond(dx));
+                this.block = m.Block;
+                m.Return();
+            });
+            ssa.Validate(s => Console.WriteLine("wut? {0}", s));
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/UnitTests/Analysis/LongAddRewriterTests.cs:528]: Fixed,

File: src/Decompiler/Evaluation/ExpressionSimplifier.cs [hunk 7]
```diff
@@ -491,9 +491,14 @@ public virtual (Expression, bool) VisitBinaryExpression(BinaryExpression binExp)
                 {
                     var c = Operator.IMul.ApplyConstants(cLeftRight, cRight);
                     if (c.IsIntegerZero)
+                    {
+                        ctx.RemoveExpressionUse(binLeft);
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Decompiler/Evaluation/ExpressionSimplifier.cs:495]: Done

File: src/UnitTests/Analysis/LongAddRewriterTests.cs [hunk 8]
```diff
@@ -419,5 +419,129 @@ public void Larw_do_not_span_multiple_blocks()
...
+                //var dx_6 = m.Reg16("dx_6");
+                //var ax_7 = m.Reg16("ax_7");
+                //var SCZO_8 = m.Flags("SCZO_8", sczo);
+                //var C_9 = m.Flags("C_9", cy);
+                //var dx_10 = m.Reg16("dx_10");
+                //var SCZO_11 = m.Flags("SCZO_11", sczo);
+
+                //m.AddDefToEntryBlock(fp);
+                //m.AddDefToEntryBlock(ax_1);
+                //m.AddDefToEntryBlock(dx_2);
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/UnitTests/Analysis/LongAddRewriterTests.cs:514]: Done

File: src/UnitTests/Analysis/ConditionCodeEliminatorTests.cs [hunk 9]
```diff
@@ -824,6 +852,174 @@ def r0
...
+                m.Assign(dx_6, m.IAdd(dx_2, C_5));
+
+                m.Assign(ax_7, m.IAdd(ax_3, m.Mem16(m.IAdd(fp, 6))));
+                m.Assign(SCZO_8, m.Cond(ax_7));
+                m.Assign(C_9, m.Slice(PrimitiveType.Bool, SCZO_8, 1));
+                m.Assign(dx_10, m.IAdd(m.IAdd(dx_6, m.Mem16(m.IAdd(fp, 8))), C_9));
+                m.Assign(SCZO_11, m.Cond(dx_10));
+                m.Return();
+            });
+            var q = @"
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/UnitTests/Analysis/ConditionCodeEliminatorTests.cs:1004]: Done

File: src/UnitTests/Analysis/LongAddRewriterTests.cs [hunk 10]
```diff
@@ -419,5 +419,129 @@ public void Larw_do_not_span_multiple_blocks()
...
+                m.Assign(ax, m.IAdd(ax, m.Mem16(m.IAdd(bx, 6))));
+                m.Assign(SCZ, m.Cond(ax));
+                //m.Alias(CF, m.Slice(PrimitiveType.Bool, SCZO, 1));
+                m.Assign(dx, m.IAdd(m.IAdd(dx, m.Mem16(m.IAdd(bx, 8))), CF));
+                m.Assign(SCZ, m.Cond(dx));
+                this.block = m.Block;
+                m.Return();
+            });
+            ssa.Validate(s => Console.WriteLine("wut? {0}", s));
+            var q = @"
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/UnitTests/Analysis/LongAddRewriterTests.cs:529]: Done

File: src/Decompiler/Analysis/ConditionCodeEliminator.cs (hunk 2 above)
Comment (uxmal) [reply src/Decompiler/Analysis/ConditionCodeEliminator.cs:495]: added a check for >= 2 uses, although these may be very unlikely.

File: src/Decompiler/Evaluation/ExpressionSimplifier.cs (hunk 1 above)
Comment (uxmal) [reply src/Decompiler/Evaluation/ExpressionSimplifier.cs:1190]: Now if the sequence contains anything else than multiplications or slices of that same multiplication, the transformation will not occur.

File: subjects/CPM-80/CB80_code.c [hunk 11]
```diff
@@ -47,7 +47,7 @@ void fn0390(byte b, Eq_n c, byte e, byte bArg02)
 bool fn03BB(Eq_n c, union Eq_n & aOut)
 {
 	g_t1657 = c;
-	byte a_n = 0x00 - (byte) (g_t1657 > 0x7B || g_t1657 < 0x61);
+	byte a_n = SLICE(0x1A - (uint16) (g_t1657 - 0x61), byte, 8);
```
[the remark this answers is not in the document]
Comment (uxmal) [reply subjects/CPM-80/CB80_code.c:50]: Looks very strange; where did that `8`  in the `SLICE` come from? Investigating.

**Comment by uxmal** [comment]
Commit ccd33d4 addresses most of the comments. I'm looking at the regression now.

**Comment by uxmal** [comment]
I've created a separate issue for the regression in CB80.

**Comment by uxmal** [comment]
Code has been committed to `master`.

**Pull request closed without merging**